compound_compat.hh

```diff
@@ -71,8 +71,9 @@ public:
 
         iterator(const legacy_compound_view& v)
             : _singular(v._type.is_singular())
-            , _offset(_singular ? 0 : -2)
-            , _i(v._type.begin(v._packed))
+            , _offset(_singular && v._type.begin(v._packed)->size() ? 0 : -2)
```

Comment:
This smells wrong. Why should we have -2 for an empty key? It yields the correct result but is not natural to have -2 offset (for the size bytes) when they won't be emitted.

I think it's cleaner to keep this 0, but adjust the end_tag constructor to set _offset to 0 for singular keys.

Reply:
It's preexisting (see lines 92-94 in this file), and moreover, it's specified like that in CQL. https://github.com/apache/cassandra/blob/ac34f28e31bbc294db80e2b5d06f8904a7f41a46/doc/native_protocol_v4.spec#L229-L233 Weird, but we need to remain compatible nonetheless

Reply:
No, this is a different -2. This -2 leaves 2 bytes space for the size of the component. See operator*().

Reply:
Oh, right, this isn't CQL. Still, operator* already had this branch - so this is a hack, but still preexisting, for the sake of some other corner case? In that case we should probably only fix that in a follow-up

Reply:
For reference, Tomek was kind enough in 2015 to leave a graphical comment about these special values:
https://github.com/scylladb/scylla/blob/f2060ac03bf008b40bc8d0ece005b79745ca593b/compound_compat.hh#L62-L67

Reply:
The -2 is fine for compound keys, but not singular keys. See my suggestion above (clarified here):

I think it's cleaner to keep this 0 _for singular keys_, but adjust the end_tag constructor to set _offset to 0 for singular keys.

Reply:
@avikivity right, that's also an option. I actually considered it but both seemed equally ugly :-( I went with my solution because it meant adding ugly conditionals only to the begin tag (which already had one such ugly conditional), and not the the end tag. But I'll prepare a new version where I uglify both - to reach a more balanced ugliness ;-)

Reply:
@avikivity I looked at the code again, and that "-2" end signal is sort of built into the code... When operator++ finishes iteration, it explicitly sets _offset = -2. It could have been anything else, but it's set to -2. That's why it makes sense for the end iterator to unconditionally return -2, and the begin() of an empty string to match it.
But that being said, 0 is ok too, because if begin==end (and that's the whole point) operator++ is never called so never sets -2 anyway. I'll prepare a patch with the 0.